File: src/dotnet10-server/RuntimeOperatingSystem.cs
namespace Dotnet10Server;

using System;
using System.IO;
using System.Runtime.InteropServices;

internal sealed record RuntimeOperatingSystem(string Name, string Source)
{
    internal static RuntimeOperatingSystem Read()
    {
        return RuntimeInformation.IsOSPlatform(OSPlatform.Linux)
            ? ReadLinux("/etc/os-release", "/usr/lib/os-release")
            : new(RuntimeInformation.OSDescription, "RuntimeInformation.OSDescription");
    }

    internal static RuntimeOperatingSystem ReadLinux(params string[] paths)
    {
        foreach (string path in paths)
        {
            string contents;
            try
            {
                contents = File.ReadAllText(path);
            }
            catch (FileNotFoundException)
            {
                continue;
            }
            catch (DirectoryNotFoundException)
            {
                continue;
            }

            return new(contents, path);
        }

        return new("unexposed-by-runtime", "unexposed-by-runtime");
    }
}

File: src/java-netty-server/src/main/java/wheretotls/RuntimeOperatingSystem.java
package wheretotls;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;

record RuntimeOperatingSystem(String name, String source) {
    static RuntimeOperatingSystem read() {
        String name = System.getProperty("os.name");
        if (!"Linux".equalsIgnoreCase(name)) {
            return new RuntimeOperatingSystem(name + " " + System.getProperty("os.version"),
                    "System.getProperty(os.name, os.version)");
        }
        try {
            return readLinux(Path.of("/etc/os-release"), Path.of("/usr/lib/os-release"));
        } catch (IOException exception) {
            throw new IllegalStateException("failed to read runtime OS evidence", exception);
        }
    }

    static RuntimeOperatingSystem readLinux(Path... paths) throws IOException {
        for (Path path : paths) {
            String contents;
            try {
                contents = Files.readString(path);
            } catch (NoSuchFileException exception) {
                continue;
            }
            return new RuntimeOperatingSystem(contents, path.toString());
        }
        return new RuntimeOperatingSystem("unexposed-by-runtime", "unexposed-by-runtime");
    }
}
